[JENKINS-37032] Add a UI to allow controlling enabled agent protocols

stephenc opened the pull request:
Follow-up to [JENKINS-37031](https://issues.jenkins-ci.org/browse/JENKINS-37031) and #2486 

See [JENKINS-37032](https://issues.jenkins-ci.org/browse/JENKINS-37032)

@jenkinsci/code-reviewers @reviewbybees

stephenc committed:
[JENKINS-37031] Advertise the enabled Agent Protocols to enable faster connection negotiation

stephenc committed:
[FIXED JENKINS-37032] Add a UI to allow controling the agent protocol enablement

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

daniel-beck commented on core/src/main/java/hudson/TcpSlaveAgentListener.java:
```diff
@@ -246,6 +261,18 @@ public PingAgentProtocol() {
+         * Allow essential {@link AgentProtocol} implementations (basically {@link PingAgentProtocol})
+         * to be always enabled.
+         *
+         * @return {@code true} if the protocol can never be disbaled.
```
typo

daniel-beck commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -1060,6 +1084,65 @@ public void setSlaveAgentPort(int port) throws IOException {
+        for (AgentProtocol p : AgentProtocol.all()) {
+            String name = p.getName();
+            if (name != null && !p.isRequired()) {
+                if (p.isOptIn()) {
```
Is this correct?

It's not clear to me what it means to be 'opt-in' -- sounds like the opposite of 'required', i.e. 'not allowed' (if false)?

stephenc replied:
Opt-in means it is disabled by default rather than enabled by default.

When we bring out JNLPProtocol4 I will start it as opt-in... Then we will do an A/B test whereby 10% of users are opt-out and 90% are opt-in... After a couple of releases with that we can switch the A/B to 50% and finally turn as opt-out permanently.

I'll probably start an A/B test to turn the legacy protocols to opt-in for 10% of users so we can reach the point where the plaintext protocols are disabled by default... But I'll do that in a separate JIRA

stephenc replied:
So when it's opt-in we record the protocol being enabled. As it is assumed disabled unless the user configures it explicitly

When it's opt-out we record the protocol not being enabled

daniel-beck replied:
So this will result in configuration changes when a protocol that used to be not opt-in (opt-out?) becomes opt-in on any given instance, _iff_ it was enabled. This appears to be reasonable, otherwise submitting the form means that it won't ever be disabled this way.

IMO not obvious behavior and there appears to be no documentation for it.

daniel-beck commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -599,6 +601,28 @@ protected void onModified() throws IOException {
+     *
+     * @since FIXME
+     */
+    private String disabledAgentProtocols;
```
Why is this stored as a string?

stephenc replied:
So that the element is not present when not configured... I'll see if I can bash Xstream into a nicer config without using a string

daniel-beck commented on core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy:
```diff
@@ -28,6 +29,20 @@ l.layout(norefresh:true, permission:app.ADMINISTER, title:my.displayName, csscla
+                        for (AgentProtocol p: AgentProtocol.all()) {
+                            if (p.name != null && !p.required) {
+                                f.checkbox(name: "agentProtocol",
+                                        title: p.name,
```
Should agent protocols have an optional display name / help file for this UI to enable users to make an actual informed decision.

stephenc replied:
A reasonable suggestion. I'll see what I can do

stephenc committed:
[JENKINS-37032] Add display names and optional descriptions

stephenc committed:
[JENKINS-37032] Switch to element per protocol XML format

- Also switch the A/B testing of JNLP3-connect to via opt-in
- Also mark that JNLP3-connect will be deprecated by JENKINS-36871
- Also start A/B testing the disabling of JNLP-connect and CLI-connect (should be well safe by now)

stephenc commented:
Showing the Advanced button when not expanded:

![screen shot 2016-07-29 at 10 18 22](https://cloud.githubusercontent.com/assets/209336/17243682/00f93a54-5576-11e6-814b-3fc14e4732c7.png)

Showing the Advanced button when expanded:

![screen shot 2016-07-29 at 10 18 29](https://cloud.githubusercontent.com/assets/209336/17243701/11c29ad8-5576-11e6-9351-5fb7dda867bf.png)

Showing the XML format with some protocols configuration toggled:

![screen shot 2016-07-29 at 10 19 01](https://cloud.githubusercontent.com/assets/209336/17243714/204f3912-5576-11e6-8b10-ce1013bcccf0.png)

@daniel-beck happy now?

stephenc commented:
Oh and if you leave everything with the defaults, this is what the config.xml looks like:
![screen shot 2016-07-29 at 10 22 01](https://cloud.githubusercontent.com/assets/209336/17243770/6014f924-5576-11e6-97df-675ca2611665.png)

daniel-beck commented:
Very happy. Thanks @stephenc 

:+1: LGTM.

stephenc committed:
[JENKINS-37032] Document the logic for the preference recording

stephenc commented:
@daniel-beck so hopefully with this comment in f02c6c9 you are now ecstatic!!! ;-)

daniel-beck commented:
@stephenc Wondering whether this would make a better Javadoc comment on `isOptIn`. Should be visible to protocol designers.

Still, :bee:

stephenc committed:
[JENKINS-37032] More documentation of the logic for the preference recording
